LineHeightControl: Allow for more granular control of decimal places

mikachan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This adds an additional `step` value to the `LineHeightControl` of `0.01`, which allows for more granular control over the line-height value.

Currently, users are only able to use values with single decimal places, such as 1.5 or 1.6. But there is quite a big difference between just one decimal place, and it would be great if smaller values could be set, such as 1.55, etc.

Please see https://github.com/WordPress/gutenberg/issues/42998 and https://github.com/WordPress/gutenberg/issues/51754 for more details.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Closes https://github.com/WordPress/gutenberg/issues/42998.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Adds a `spinFactor` prop to the NumberControl component to opt-in for more granular control over decimal places.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
1. Go to the Post or Site editor.
2. Insert a Paragraph block or any block that supports line-height.
3. Try changing the line-height using the block controls to a value with two decimal places, e.g. 1.62.
4. The value should not automatically round up or down to a value with one decimal place and should remain the same value with two decimal places.
5. Using the arrow keys or the spin controls, increase and decrease the line-height. The value should change by 0.1.

## Screenshots or screencast <!-- if applicable -->

https://github.com/WordPress/gutenberg/assets/1645628/b6f4c78e-9861-4461-bc2c-94ab5ed834eb

## Dev note

A new `spinFactor` prop has been added to `NumberControl`, allowing consumers of the components to specify by how much should the input's value increment/decrement with respect to the `step` prop. This is particularly useful when more granular control is needed (thus allowing for more decimal places and a smaller `step`) but without sacrificing the UX around manually incrementing/decrementing the value.

Even if the prop was added to `NumberControl`, all components based on `NumberControl` can benefit from this change — this includes, for example, `UnitControl` and `LineHeightControl`.

mikachan committed:
Change the line height step value from 0.1 to 0.01

mikachan committed:
Update test and related comments

richtabor commented:
I don't think this is a good idea.

Pressing 10x more times to get + 0.1 point change doesn't seem ideal. 

The current implementation is `0.1` stepped, which provides enough of a difference in a short enough process, if you're using the + and - controls. If you want something more specific, it's easier to type in, say `1.55` instead of pressing the + five times (assuming you start at 1.5). 

And if you want to go from 1.5 to 1.7, it's quite a heavy UX lift, whereas today it's only two clicks.

hanneslsm commented:
> Pressing 10x more times to get + 0.1 point change doesn't seem ideal.
> 
> The current implementation is `0.1` stepped, which provides enough of a difference in a short enough process, if you're using the + and - controls. If you want something more specific, it's easier to type in, say `1.55` instead of pressing the + five times (assuming you start at 1.5).
> 
> And if you want to go from 1.5 to 1.7, it's quite a heavy UX lift, whereas today it's only two clicks.

I was about to comment the same.
While I absolutely think that it should be possible to have line-heights of `1.23`, I think it's an issue of the input element
From what I can tell, the input with + and - is a custom one, @richtabor do you know when and why it was introduced?
We also do have other situations where 10-step is only possible but another input field is used.  (https://github.com/WordPress/gutenberg/issues/51625)

I think the best solution would be to make it possible to enter values `1.23` manually, but have the plus and minus to jump/reset the value to `1.3` or `1.2`

mikachan committed:
Revert 0.1 to 0.01 step changes

mikachan committed:
Avoid using roundClamp for line height calculations

mikachan committed:
Introduce roundingStep prop to NumberControl

mikachan committed:
Update number-control tests to use string input values

mikachan committed:
Add changelog entry

mikachan committed:
Merge branch 'trunk' into update/line-height-step

mikachan commented:
Thanks, both, for the feedback! I agree; whilst that solution allowed for 2 decimal places, it introduced a UX regression.

I've updated the PR with another idea, which allows the user to manually enter up to 2 decimal places but will also maintain the step value of `0.1` when using the spin controls or arrow keys. I've updated the screencast in the video so you can see how it works.

The value in the `NumberControl` line-height input is calculated using `roundClamp()`, which rounds the input value based on the `step` value. In order to allow both the `0.1` step input controls and the `0.01` for manually adding 2 decimal places, the `0.01` step value needs to be passed to `roundClamp`, otherwise this function removes the second decimal place (as it's rounding the value based on `0.1`).

So I've introduced an optional prop that can be passed to the `NumberControl` component called `roundingStep`, which is used instead of the default `step` value if it exists, and is only used in the rounding calculation. I tried to use a solution that didn't interfere too much with the existing calculations but also wasn't specific to the line-height control, so other components using the `NumberControl` can use it where needed.

hanneslsm commented:
Thanks @mikachan !
From what I can tell from the video, after entering `1.25` and clicking `-` once I'd end up at `1.15`?
Not sure if this is a common usage. I think it'd be better to click once `-` and get back to a rounded value, meaning `1.2`. This way I can quickly test if `1.25` (*manually typed in*) `1.2` (*click*) or `1.0` (*click* *click*) works best.

For context, for most font sizes it's common to have rounded values. However, sometimes it's needed to have the second decimal number. That's what this PR is for. For the input field controls, I'd go with the rounding. Examples: https://typescale.com/

richtabor commented:
> We also do have other situations where 10-step is only possible but another input field is used. (https://github.com/WordPress/gutenberg/issues/51625)

This is a limitation because of how the block was originally introduced; it's manually applying CSS classes for every tenth value. 😅

> I've updated the PR with another idea, which allows the user to manually enter up to 2 decimal places but will also maintain the step value of 0.1 when using the spin controls or arrow keys. I've updated the screencast in the video so you can see how it works.

Looks ace!

richtabor reviewed (commented):
Design + UX wise this feels much better than what we have in trunk. 🚀

stokesman reviewed (commented):
Thank you for taking this up 🙌  The current changes do resolve the issue. While it doesn't cause practical problems that I'm aware of, this part bugs me:

<img width="275" alt="image" src="https://github.com/WordPress/gutenberg/assets/9000376/98f4194d-5fdc-49fe-b9e6-22b4d7e8473e">

That's Chrome on macOS if you hover the input. This is due to `step` being more a validation attribute than anything else.

Because of that, I think your first pass of setting the line height control’s `step` to `.01` was the way to go. Since the granularity of the stepping UX should be preserved, we could introduce a prop to specify that instead. I'd feel better about that direction. WDYT?

<details><summary>A diff of quick and dirty changes toward suggested direction</summary>

```diff
diff --git a/packages/block-editor/src/components/line-height-control/index.js b/packages/block-editor/src/components/line-height-control/index.js
index dc0cf05d18..62233c9d93 100644
--- a/packages/block-editor/src/components/line-height-control/index.js
+++ b/packages/block-editor/src/components/line-height-control/index.js
@@ -111,8 +111,8 @@ const LineHeightControl = ( {
 				onChange={ handleOnChange }
 				label={ __( 'Line height' ) }
 				placeholder={ BASE_DEFAULT_VALUE }
-				step={ STEP }
-				roundingStep={ ROUNDING_STEP }
+				step={ ROUNDING_STEP }
+				spincrement={ STEP }
 				value={ value }
 				min={ 0 }
 				spinControls="custom"
diff --git a/packages/components/src/number-control/index.tsx b/packages/components/src/number-control/index.tsx
index 1909937b97..bbb900ec4e 100644
--- a/packages/components/src/number-control/index.tsx
+++ b/packages/components/src/number-control/index.tsx
@@ -43,7 +43,7 @@ function UnforwardedNumberControl(
 		required = false,
 		shiftStep = 10,
 		step = 1,
-		roundingStep,
+		spincrement = step,
 		type: typeProp = 'number',
 		value: valueProp,
 		size = 'default',
@@ -65,7 +65,8 @@ function UnforwardedNumberControl(
 	const mergedRef = useMergeRefs( [ inputRef, forwardedRef ] );
 
 	const isStepAny = step === 'any';
-	const baseStep = isStepAny ? 1 : ensureNumber( step );
+	let baseStep = ensureNumber( spincrement );
+	if ( spincrement === step ) baseStep = isStepAny ? 1 : ensureNumber( step );
 	const baseValue = roundClamp( 0, min, max, baseStep );
 	const constrainValue = (
 		value: number | string,
@@ -73,15 +74,11 @@ function UnforwardedNumberControl(
 	): string => {
 		// When step is "any" clamp the value, otherwise round and clamp it.
 		// Use '' + to convert to string for use in input value attribute.
-		return isStepAny
-			? '' + Math.min( max, Math.max( min, ensureNumber( value ) ) )
-			: '' +
-					roundClamp(
-						value,
-						min,
-						max,
-						stepOverride ?? ( roundingStep || baseStep )
-					);
+		return `${
+			isStepAny
+				? Math.min( max, Math.max( min, ensureNumber( value ) ) )
+				: roundClamp( value, min, max, stepOverride ?? baseStep )
+		}`;
 	};
 
 	const autoComplete = typeProp === 'number' ? 'off' : undefined;

```
</details>

richtabor commented:
> Because of that, I think your first pass of setting the line height control’s step to .01 was the way to go. Since the granularity of the stepping UX should be preserved, we could introduce a prop to specify that instead. I'd feel better about that direction. WDYT?

I mentioned earlier why .01 stepping holistically would make line height + and - quite tedious. I wouldn’t want to press + numerous times to really modify the line height.

stokesman commented:
My bad if I didn't make that very clear but in part of what you quoted:

> Since the granularity of the stepping UX should be preserved, we could introduce a prop to specify that instead

Meaning that the direction I've suggested maintains the current UX.

mikachan committed:
Try spincrement prop

chad1008 reviewed (commented):
Thanks for working on this, it feels like a great change! I agree with what others have shared  regarding the UX, and I think @stokesman's diff feels pretty clean in terms of having the spin controls/arrow keys round to the value back to the step amount when clicked.

There are some rounding issues to address there (I'm not criticizing the diff, I know it was a quick demo 😄). For example, if you apply the diff and set the line height to `1.45`, hitting `+`/up arrow will jump you all the way to `1.6`. Similarly, hitting `-`/down arrow from `1.44` jumps straight to `1.3`.

It looks like it's rounding to the nearest step when clicked and then incrementing/decrementing, when we'd really just want it to round up/down whenever there's a second decimal in place.

mikachan committed:
Handle rounding logic with two decimals

mikachan commented on packages/components/src/number-control/index.tsx:
```diff
@@ -120,7 +123,6 @@ function UnforwardedNumberControl(
 				type === inputControlActionTypes.PRESS_UP ||
 				type === inputControlActionTypes.PRESS_DOWN
 			) {
-				// @ts-expect-error TODO: Resolve discrepancy between `value` types in InputControl based components
```
Just to clarify this only-slightly related change:

The discrepancy here was caused by `constrainValue()` either returning a number or a string and in this PR I've updated this function so that it only returns a string (see [line 73](https://github.com/WordPress/gutenberg/pull/52902/files#diff-1795d70ae6f25ef5ee79fa75d215a18e87de7d7bdc577b294376838325d983a8R73)), which is why these `@ts-expect-error`s are no longer needed. I believe a string is all that is needed, as input values are always strings.

mikachan committed:
Merge branch 'trunk' into update/line-height-step

mikachan committed:
Fix rounding up logic

mikachan committed:
Merge branch 'trunk' into update/line-height-step

mikachan committed:
Tweak type comments and changelog entry

Co-Authored-By: Mitchell Austin <9000376+stokesman@users.noreply.github.com>

mikachan committed:
Merge branch 'trunk' into update/line-height-step

mikachan commented:
Thanks all, for all the feedback and input! 🙇

I really like the `spincrement` diff from @stokesman, so I've applied the ideas from that diff and then added some additional logic to handle the rounding calculations mentioned by @chad1008. I used `roundClamp` to round the line-height based on the `STEP` value rather than the `ROUNDING_STEP` value when there are two decimal places. This means that if a value ends in .x5, incrementing the value will round up to the next 0.1 value, and decrementing the value will round down to the current 0.1 value.

I think this is ready for another review!

stokesman commented:
Thanks for the updates Sarah!

> added some additional logic to handle the rounding calculations mentioned by @chad1008.

I agree with fixing the issue Chad pointed out and ideally doing so in `NumberControl` so that it’s not an issue elsewhere. However, I don't think the rounding should be coarser than the `step`. My expectation is that from `1.45`, hitting `+`/up arrow results in `1.55`. 

<details><summary>A diff fixing the spin issue in <code>NumberControl</code></summary>

```diff
diff --git a/packages/components/src/number-control/index.tsx b/packages/components/src/number-control/index.tsx
index 6526b93694..678b88d62f 100644
--- a/packages/components/src/number-control/index.tsx
+++ b/packages/components/src/number-control/index.tsx
@@ -65,8 +65,9 @@ function UnforwardedNumberControl(
 	const mergedRef = useMergeRefs( [ inputRef, forwardedRef ] );
 
 	const isStepAny = step === 'any';
-	let baseStep = ensureNumber( spincrement );
-	if ( spincrement === step ) baseStep = isStepAny ? 1 : ensureNumber( step );
+	const [ baseSpin, baseStep ] = isStepAny
+		? [ 1, 1 ]
+		: [ ensureNumber( spincrement ), ensureNumber( step ) ];
 	const baseValue = roundClamp( 0, min, max, baseStep );
 	const constrainValue = (
 		value: number | string,
@@ -91,7 +92,7 @@ function UnforwardedNumberControl(
 	) => {
 		event?.preventDefault();
 		const shift = event?.shiftKey && isShiftStepEnabled;
-		const delta = shift ? ensureNumber( shiftStep ) * baseStep : baseStep;
+		const delta = shift ? ensureNumber( shiftStep ) * baseSpin : baseSpin;
 		let nextValue = isValueEmpty( value ) ? baseValue : value;
 		if ( direction === 'up' ) {
 			nextValue = add( nextValue, delta );
@@ -137,8 +138,8 @@ function UnforwardedNumberControl(
 				const [ x, y ] = payload.delta;
 				const enableShift = payload.shiftKey && isShiftStepEnabled;
 				const modifier = enableShift
-					? ensureNumber( shiftStep ) * baseStep
-					: baseStep;
+					? ensureNumber( shiftStep ) * baseSpin
+					: baseSpin;
 
 				let directionModifier;
 				let delta;

```
</details>

If you apply the diff as is, the line height control is still going to round off the hundredth so to test you'd have to either try a similar setup of `NumberControl` in Storybook or revert the changes for the special rounding in the line height control.

--- 

I also realized a more general concern about the `spincrement` prop. It seems a bad idea to implement as an amount independent of `step` because it sets up more potential to create inputs that "spin" to invalid values. I think I should have proposed it as a multiplier of `step` instead. That way a spin should always land on a valid step. I've push a [branch based on this one to demonstrate](https://github.com/WordPress/gutenberg/compare/update/line-height-step...update/line-height-step-w-spinFactor?expand=1) the idea. In that branch I reverted the special rounding of the line height control because I think the fixed baseline in `NumberControl` suffices and to make it easy to try it that way.

hanneslsm commented:
@stokesman 
> My expectation is that from 1.45, hitting +/up arrow results in 1.55.

In general you're right, but in the context of line-height it does not make too much sense. See my comment from above: https://github.com/WordPress/gutenberg/pull/52902#issuecomment-1653381172

stokesman commented:
Hannes, thanks for bringing that up again. I did see and consider your earlier comment yet it seems quite arbitrary. I.e. it's not clear which behavior is favorable in common use.

> This way I can quickly test if `1.25` (_manually typed in_) `1.2` (_click_) or `1.0` (_click_ _click_) works best.

As example of why this seems arbitrary, this alternative way is as quick or quicker:
I can quickly test if `1.25` (_manually typed in_) `1.2` (_backspace_) or `1.0` (<kbd>↓</kbd><kbd>↓</kbd>) works best.

I’m not really opposed to it but I think we might hold off on specializing it that way for the moment.

hanneslsm commented:
@stokesman 
I posted my comment and reevaluated shortly after. I deleted the comment but had to run before I could finish my new answer. Sorry.

I think the biggest challenge on deciding how the component should behave is that the plus and minus are visually very prominent and the element is not not known for it's behaviour yet.
Usually, we'd have an input field with arrows and they would probably in- & decrease the value by 0.01. As already discussed, not ideal here. The plus and minus definitely has its advantages. 

> I’m not really opposed to it but I think we might hold off on specializing it that way for the moment.

I agree and can see that people expect them to change the value by 0.1 (meaning `1.15` will become `1.25`) no matter what the original value is. Actually, I am leaning more to agreeing with you and (for now) not specialising the input field even more. 

> I can quickly test if 1.25 (manually typed in) 1.2 (backspace) or 1.0 (↓↓) works best.

This is a very valid point emphasises the argument.

I guess I have change my mind and agree with you that we should (for now) avoid a special rounding. 
Let's see how the users react.

mikachan committed:
Revert specialised rounding in LineHeightControl

mikachan committed:
Fix spin rounding in NumberControl

Co-Authored-By: Mitchell Austin <9000376+stokesman@users.noreply.github.com>

mikachan committed:
Replace spincrement with spinFactor

Co-Authored-By: Mitchell Austin <9000376+stokesman@users.noreply.github.com>

mikachan commented:
Thanks both! I really appreciate the speedy feedback.

> I agree and can see that people expect them to change the value by 0.1 (meaning 1.15 will become 1.25) no matter what the original value is.

I'm not too opinionated either way, but I also agree here, and it's potentially complicating things by specialising the logic, at least for now.

I've tried out both of @stokesman's diffs from the above [comment](https://github.com/WordPress/gutenberg/pull/52902#issuecomment-1668431059), and I really like how this is working. I agree that the new prop (`spinFactor`) works better as a multiplier of `step`. I also agree with fixing the rounding logic directly in `NumberControl` - I tried out a few things with my previous attempt and decided it was specific to line-height values, but this iteration along with the `spinFactor` prop works very nicely within `NumberControl` ✨ @stokesman, thank you for providing such specific technical feedback!

stokesman committed:
Update line height control unit tests

stokesman reviewed (approved):
Thanks again for the updates! I've taken it for another 🙊spin and everything behaves as expected. I noticed there were some unit tests that failed and manually tested to verify the behaviors still work. It was merely an implementation detail in the tests that needed an update. I pushed a commit for that so the tests should pass now.

Maybe we want sign off from one of the components crew but this looks good to go. I ❤️  the feature being added to `NumberControl` so the finer UX can be supported. Thanks again for tackling this Sarah 🙌 .

mikachan committed:
Merge branch 'trunk' into update/line-height-step

mikachan commented:
Thanks for updating that unit test! I stepped away from my computer before the tests finished yesterday and didn't run them locally 🙈 Thanks also for all your help with this PR!

mikachan commented:
@mirka Would you be able to review this please, as someone who knows the components well? Or suggest a good person for a review. Thank you 🙇

mirka commented on packages/components/src/number-control/index.tsx:
```diff
@@ -64,16 +65,22 @@ function UnforwardedNumberControl(
 
 	const isStepAny = step === 'any';
-	const baseStep = isStepAny ? 1 : ensureNumber( step );
+	let baseStep = ensureNumber( step );
+	let baseSpin = ensureNumber( spinFactor ) * baseStep;
+	if ( isStepAny ) {
+		baseStep = 1;
+		baseSpin = 1 * ensureNumber( spinFactor );
+	}
```
```suggestion
	const baseStep = isStepAny ? 1 : ensureNumber( step );
	const baseSpin = ensureNumber( spinFactor ) * baseStep;
```

Nit: It seems like we can simplify and avoid the `let`s?

mirka commented on packages/components/src/number-control/index.tsx:
```diff
@@ -64,16 +65,22 @@ function UnforwardedNumberControl(
 		value: number | string,
 		stepOverride?: number
-	) => {
+	): string => {
```
```suggestion
	) => {
```

Nit: I think we can let TypeScript do the inference here.

mikachan replied:
Oh yes, looks like we can! I thought this was helping solve the need for [these](https://github.com/WordPress/gutenberg/pull/52902/files#diff-1795d70ae6f25ef5ee79fa75d215a18e87de7d7bdc577b294376838325d983a8L123) @ts-expect-error comments, but it doesn't look like we need to specify the return type at all.

mirka reviewed (commented):
Loved seeing the collaboration here 🚀 

To be honest though I'm a little confused as to what the ultimate change is supposed to be here 🤔 This is the behavior I'm seeing, and it seems to contradict what's written in the code comments ("A line-height value of 1.55 will increment to 1.65"):

https://github.com/WordPress/gutenberg/assets/555336/1e0031bd-6def-452f-b2ad-da55272a28ff

What exactly is _supposed_ to happen after the changes in this PR?

In that vein, I think it'd be good to add some tests for the new behavior as well.

stokesman commented:
> This is the behavior I'm seeing, and it seems to contradict what's written in the code comments…

I just double-checked and don't see what you are seeing. Perhaps there was a build hiccup of some sort. This is what I'm seeing:

https://github.com/WordPress/gutenberg/assets/9000376/d6181c24-7634-4d2b-b91c-04a6f794e625

> What exactly is _supposed_ to happen after the changes in this PR?

Simply that the line height control accepts numbers with a hundredth decimal place and its behavior remains otherwise unchanged. It already "spins" by tenths and showing that it still does that is proof of preserved UX.

> I think it'd be good to add some tests for the new behavior as well.

That's a fine idea. I'm not sure we'd need one for `LineHeightControl`’s acceptance of hundredth place values, WDYT? I think we’d mostly want something covering that the new `spinFactor` prop of `NumberControl` works as intended.

mikachan committed:
Update packages/components/src/number-control/index.tsx

Co-authored-by: Lena Morita <lena@jaguchi.com>

mikachan committed:
Update packages/components/src/number-control/index.tsx

Co-authored-by: Lena Morita <lena@jaguchi.com>

mikachan committed:
Add tests that include use of spinFactor prop

mikachan committed:
Add tests for shiftStep + spinFactor

mikachan commented:
Thanks @mirka! To confirm, I'm seeing the same as @stokesman on this branch, and the biggest difference is that users can now manually add and save a line-height value with a hundredth decimal place. Currently, this decimal place is removed on blur or when interacting with the spin controls. Here's an example from trunk to compare to @stokesman's video:


https://github.com/WordPress/gutenberg/assets/1645628/baedcffb-e2de-415f-a19e-33979519d6df


I've added some tests for the `spinFactor` prop, following the patterns already used in the `NumberControl` tests. Everything worked as expected, but I'm not sure if I've added the correct number of tests (too many/too few). They're in these commits: https://github.com/WordPress/gutenberg/pull/52902/commits/4c7d5452386965e0fb360b5fa24a5811a1e7f5d1 & https://github.com/WordPress/gutenberg/pull/52902/commits/facc66f7bb40014c24beb8b1a22b2ac5bb7af252.

mikachan committed:
Merge branch 'trunk' into update/line-height-step

mirka reviewed (approved):
Thank you both for clarifying! I got it working as expected now, it must've been a build hiccup indeed. The tests look good — I think it nicely captures how the feature is supposed to work.

🚀

mikachan commented:
Woop, thanks @mirka! I'll bring this in 🎉

(I had loads of fun working on this, thanks everyone!)

ciampo commented:
Added a dev note in the PR description
